### ISSUE TITLE
Focus indicator is visible.

### ISSUE THREAD

**Issue by ecureuill** [solex2006/SELIProject#116]
This is a "not-end" requirement. What I want to mean is that this should be considered for any new page you create that Student can access. Also should correct the old pages.

I will use the label **Feature Design Notes** to this cases
 
***************
# Scenario: Focus is visible (minimum)
Components must provide a visible indication when they have the input focus.

### Examples
- When text fields receive focus, a vertical bar is displayed in the field, indicating that the user can insert text, OR all of the text is highlighted, indicating that the user can type over the text.
- When a user interface control receives focus, a visible border is displayed around it.

- [ ] Using user interface components that are **highlighted by the user agent** when they receive focus.  It is common for user agents to highlight standard controls in some way when they receive focus.  **When authors use standard controls** for which the user agent provides this support, users are informed of the focus location in a standard, predictable way.
- [ ] Using CSS to change the presentation of a user interface component when it receives focus [Live-demo](https://www.w3.org/WAI/WCAG21/working-examples/focus-pseudo-class/)
- [ ] Using script to change the background color or border of the element with focus
- [ ] Using the **default focus indicator** for the platform so that high visibility default focus indicators will carry over
> Operating systems have a native indication of focus, which is available in many user agents. The default rendering of the focus indicator isn't always highly visible and may even be difficult to see against certain backgrounds. However, many platforms allow the user to customize the rendering of this focus indicator. Assistive technology can also change the appearance of the native focus indicator. If you use the native focus indicator, any system-wide settings for its visibility will carry over to the Web page. If you draw your own focus indicator, for example by coloring sections of the page in response to user action, these settings will not carry over, and AT will not usually be able to find your focus indicator.
- [ ] **Best** Using an author-supplied, highly visible focus indicator
- [ ] **Best** Creating a two-color focus indicator to ensure sufficient contrast with all components [Live-demo 1](https://www.w3.org/WAI/WCAG21/working-examples/css-two-focus-colors/dotted.html) - [Live-demo 2](https://www.w3.org/WAI/WCAG21/working-examples/css-two-focus-colors/lined.html)

## Tests Procedure
For each element able to attain focus:
1. Using a mouse, hover over the element.
     1. Check that the background or border changes color.
1. Move the mouse away from the object before attempting keyboard focus.
1. Using a keyboard, tab to the element.
     1. Check that the background or border changes color.
1. Check that the background or border changes in color are removed when the element loses focus.

### user agent
For each focusable component in the Web page:
1. Set focus to the control
1. Check whether the user agent has highlighted the control in some way

Expected Results: 2 is true

### author-supplied, highly visible focus indicator
1. Place focus on each focusable user interface element on the page using the mouse.
1. Check that there is a highly visible focus indicator.
Place focus on each focusable user interface element on the page using the keyboard.
1. Check that the focus indicator area is at least the size of a 2 CSS px line along the longest edge of the component.
1. Check that the change of contrast of the indicator between focused and unfocused states has a ratio of 3:1 for the minimum focus indicator area.
1. If the focus indicator does not have 3:1 contrast ratio with its adjacent colors, check that it is at least 2px thick.

Expected Results: Checks 2, 4, 5 and 6 are true.

### two-color
For each focusable user interface element:
1. Check that the two-colors in the focus indicator are adjacent and have a contrast ratio that is 3:1 or greater with each other.
1. Check that the colors used contrast with the adjacent colors, or are at least 2 CSS pixels thick.

Expected Results: all are true
# Scenario: Focus Visible (Enhanced)

When a User Interface Component displays a visible keyboard focus, all of the following are true:

- Minimum area: The focus indication area is greater than or equal to the longest side of the bounding rectangle of the focused control, times 2 CSS pixels.
- Focus contrast: Color changes used to indicate focus have at least a 3:1 contrast ratio with the colors changed from the unfocused control.
- Contrast or thickness: The focus indication area has a 3:1 contrast ratio against all adjacent colors for the minimum area or greater, or has a thickness of at least 2 CSS pixels.

### 1. Minimum area

The bigger the visible change the easier it is for someone to see, authors are encouraged to make the change as large as possible. For example, a thick outline around the element.

![image](https://user-images.githubusercontent.com/993369/80006557-6820d180-849b-11ea-8fa3-8bc78bbff655.png)

To establish if a focus indicator has a size equivalent to at least a 2px line along the longest side: If a link is 90px wide by 30px tall, the indicator would need to be have a minimum area of 180px square. (I.e. 2px wide along the longest side is 2 x 90px.)

The following 3 examples use a 90px wide by 30px tall button, therefore the surface area requirement is 180px. The middle button is focused in each example.

![Figure 2 Three dark blue buttons on a white background. The middle button has a yellow inner border.; Figure 3 Three dark blue buttons on a white background. The middle button has a thick yellow line under the text.; Figure 3 Three dark blue buttons on a white background. The middle button has a yellow square to the left of the text.](https://user-images.githubusercontent.com/993369/80006581-7373fd00-849b-11ea-8b7f-84e75c7839c6.png)

**Typical focus indicators:**
- An outline around the whole component would pass the size requirement;
- Changing the background of a control would pass the size requirement;
- A 1px wide vertical line (such as a blinking cursor) would not pass the size requirement, a text input would need a larger or separate focus indicator.

**Unusual shapes and gradients**
If you have an unusal shape, some mathematics may be required if the focus indicator is on the edge of the size requirement. For example, if the focusable control is a circle with a diameter of 100px, the bounding box of that component would be a square of 100px on each side. If the focus indicator is also circular, then the area needs to 100px * 2 = 200px.

![Two orange circlular buttons with an icon in the middle. The right hand circle has a dark outline around the whole cicle to indicate focus](https://www.w3.org/WAI/WCAG21/Understanding/img/focus-indicator-circle.png)
Figure 5A circular button in the default state, and then with a focus indicator.

The circumference of a 100px diameter circle would be 314px (diameter * π), therefore a 1px outline around the cicle will be 314px of surface area, meeting the requirement for a 200px surface area.

If a focus indicator has a gradient, the principle is to measure the contrast of the changed area, and ignore the change that does not me 3:1.
![Three buttons, the middle with a heavy drop-shadow indicating focus.
](https://www.w3.org/WAI/WCAG21/Understanding/img/focus-indicator-box-shadow.png)
Figure 7When a gradient is used on a focus indictor, the measure of surface area should only include the area that has changed enough to meet the 3:1 contrast ratio.

If you take some spot-checks on the gradient area and establish what area meets the contrast, it is straightforward to work out if that area is sufficient.

![The middle button with the drop-shadow included, but the subtle grey areas removed to only show the contrasting area.](https://www.w3.org/WAI/WCAG21/Understanding/img/focus-indicator-box-shadow-only.png)
Figure 8The focused button with the non-contrasting areas removed, showing that it is a thick indicator that meets the requirements.

### 2. Contrast or thickness

When a component changes to include a focus indicator, that change can always be measured as a change of color contrast. For example, if a yellow outline is added to a button on a blue background, the change of color is from blue to yellow.

![image](https://user-images.githubusercontent.com/993369/80008450-2a717800-849e-11ea-82ef-8b06a194e3c3.png)

# Tests  Procedure

### Creating a focus indicator within the component
For each focusable user interface element:
1. Place keyboard focus on each focusable user interface element on the page using the keyboard.
2. Check that the focus indicator area is at least the size of a 2 CSS px line along the longest edge of the component.
3. Check that the change of contrast of the indicator between focused and unfocused states has a ratio of 3:1 for the minimum focus indicator area.
4. If the focus indicator does not have 3:1 contrast ratio with its adjacent colors, check that it is at least 2px thick.

Expected Results 2, 3, 4 are true.
